File: axiomgraph/routing.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List

from .graph import CognitiveEdge, CognitiveGraph


@dataclass(frozen=True)
class RouteStep:
    from_node: str
    to_node: str
    score: float


class Router:
    """Scores possible cognitive routes."""

    def __init__(self, graph: CognitiveGraph) -> None:
        self.graph = graph

    @staticmethod
    def edge_score(edge: CognitiveEdge, task_relevance: float = 1.0) -> float:
        numerator = edge.weight * edge.trust_score * max(0.01, edge.success_rate or 0.5) * task_relevance
        denominator = max(0.01, edge.token_cost * edge.latency_ms * edge.risk_score)
        return numerator / denominator

    def best_next_steps(self, from_node: str, limit: int = 3) -> List[RouteStep]:
        scored = [RouteStep(edge.from_node, edge.to_node, self.edge_score(edge)) for edge in self.graph.outgoing(from_node)]
        return sorted(scored, key=lambda step: step.score, reverse=True)[:limit]

File: hosted-mcp/src/verdigraph/routing.ts
// src/verdigraph/routing.ts — TS port of verdigraph/routing.py.

import type { CognitiveEdge, CognitiveGraph } from "./graph";
import { edgeSuccessRate } from "./graph";

export interface RouteStep {
  from_node: string;
  to_node: string;
  score: number;
}

export class Router {
  constructor(public graph: CognitiveGraph) {}

  static edgeScore(edge: CognitiveEdge, taskRelevance = 1.0): number {
    const sr = edgeSuccessRate(edge);
    const numerator = edge.weight * edge.trust_score * Math.max(0.01, sr || 0.5) * taskRelevance;
    const denominator = Math.max(0.01, edge.token_cost * edge.latency_ms * edge.risk_score);
    return numerator / denominator;
  }

  bestNextSteps(fromNode: string, limit = 3): RouteStep[] {
    const scored: RouteStep[] = this.graph.outgoing(fromNode).map((e) => ({
      from_node: e.from_node,
      to_node:   e.to_node,
      score:     Router.edgeScore(e),
    }));
    scored.sort((a, b) => b.score - a.score);
    return scored.slice(0, limit);
  }
}
